File: restim-maui/Models/SettingsModels.cs
using System;
using System.Collections.Generic;

namespace RestimMaui.Models
{
    public enum DeviceType { None, AudioThreePhase, FocStimThreePhase, NeoStimThreePhase, FocStimFourPhase }
    public enum WaveformType { Continuous, PulseBased, ABTesting }

    public class DeviceSettings
    {
        public DeviceType DeviceType { get; set; } = DeviceType.FocStimThreePhase;
        public WaveformType WaveformType { get; set; } = WaveformType.Continuous;
        public float MinFrequency { get; set; } = 500;
        public float MaxFrequency { get; set; } = 1500;
        public float WaveformAmplitude { get; set; } = 0.120f;
        public float Calibration3Center { get; set; } = -0.5f;
        public float Calibration3Up { get; set; } = 0f;
        public float Calibration3Left { get; set; } = 0f;
    }

    public class PulseSettings
    {
        public float CarrierFrequency { get; set; } = 700;
        public float PulseFrequency { get; set; } = 50;
        public float PulseWidth { get; set; } = 5;
        public float PulseRiseTime { get; set; } = 3;
        public float PulseIntervalRandom { get; set; } = 10;
    }

    public class MediaSyncSettings
    {
        public bool HereSphereEnabled { get; set; }
        public string HereSphereIp { get; set; } = "";
        public int HereSpherePort { get; set; } = 23554;
        public List<FunscriptLocation> FunscriptLocations { get; set; } = new List<FunscriptLocation>();
    }

    public class FunscriptLocation
    {
        public string Id { get; set; } = Guid.NewGuid().ToString();
        public string Name { get; set; } = "";
        public string Type { get; set; } = "local"; // local, webdav
        public bool Enabled { get; set; } = true;
        public string LocalPath { get; set; } = "";
        public string WebDavUrl { get; set; } = "";
        public string Username { get; set; } = "";
        public string Password { get; set; } = "";
    }
    
    public class FocStimSettings
    {
        public string WifiIp { get; set; } = "192.168.1.1";
        public string SerialPort { get; set; } = "COM3";
    }
}

File: restim-flutter/lib/models/settings_models.dart
class DeviceSettings {
  double minFrequency = 500;
  double maxFrequency = 1500;
  double waveformAmplitude = 0.120; // 120mA
  
  // Calibration
  double calibration3Center = -0.5;
  double calibration3Up = 0.0;
  double calibration3Left = 0.0;

  Map<String, dynamic> toJson() => {
        'minFrequency': minFrequency,
        'maxFrequency': maxFrequency,
        'waveformAmplitude': waveformAmplitude,
        'calibration3Center': calibration3Center,
        'calibration3Up': calibration3Up,
        'calibration3Left': calibration3Left,
      };

  DeviceSettings.fromJson(Map<String, dynamic> json) {
    minFrequency = json['minFrequency'] ?? 500;
    maxFrequency = json['maxFrequency'] ?? 1500;
    waveformAmplitude = json['waveformAmplitude'] ?? 0.120;
    calibration3Center = json['calibration3Center'] ?? -0.5;
    calibration3Up = json['calibration3Up'] ?? 0.0;
    calibration3Left = json['calibration3Left'] ?? 0.0;
  }
  
  DeviceSettings();
}

class PulseSettings {
  double carrierFrequency = 700;
  double pulseFrequency = 50;
  double pulseWidth = 5;
  double pulseRiseTime = 3;
  double pulseIntervalRandom = 10;

  Map<String, dynamic> toJson() => {
        'carrierFrequency': carrierFrequency,
        'pulseFrequency': pulseFrequency,
        'pulseWidth': pulseWidth,
        'pulseRiseTime': pulseRiseTime,
        'pulseIntervalRandom': pulseIntervalRandom,
      };

  PulseSettings.fromJson(Map<String, dynamic> json) {
    carrierFrequency = json['carrierFrequency'] ?? 700;
    pulseFrequency = json['pulseFrequency'] ?? 50;
    pulseWidth = json['pulseWidth'] ?? 5;
    pulseRiseTime = json['pulseRiseTime'] ?? 3;
    pulseIntervalRandom = json['pulseIntervalRandom'] ?? 10;
  }
  
  PulseSettings();
}

class FocStimSettings {
  String wifiIp = "192.168.1.1";
  int wifiPort = 55533;

  Map<String, dynamic> toJson() => {
    'wifiIp': wifiIp,
    'wifiPort': wifiPort,
  };

  FocStimSettings.fromJson(Map<String, dynamic> json) {
    wifiIp = json['wifiIp'] ?? "192.168.1.1";
    wifiPort = json['wifiPort'] ?? 55533;
  }
  
  FocStimSettings();
}

class MediaSyncSettings {
  bool hereSphereEnabled = false;
  String hereSphereIp = "";
  int hereSpherePort = 23554;
  List<FunscriptLocation> funscriptLocations = [];

  Map<String, dynamic> toJson() => {
    'hereSphereEnabled': hereSphereEnabled,
    'hereSphereIp': hereSphereIp,
    'hereSpherePort': hereSpherePort,
    'funscriptLocations': funscriptLocations.map((e) => e.toJson()).toList(),
  };

  MediaSyncSettings.fromJson(Map<String, dynamic> json) {
    hereSphereEnabled = json['hereSphereEnabled'] ?? false;
    hereSphereIp = json['hereSphereIp'] ?? "";
    hereSpherePort = json['hereSpherePort'] ?? 23554;
    if (json['funscriptLocations'] != null) {
      funscriptLocations = (json['funscriptLocations'] as List)
          .map((e) => FunscriptLocation.fromJson(e))
          .toList();
    }
  }

  MediaSyncSettings();
}

class FunscriptLocation {
  String id = DateTime.now().millisecondsSinceEpoch.toString();
  String name = "";
  String type = "local";
  String localPath = "";

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'type': type,
    'localPath': localPath,
  };

  FunscriptLocation.fromJson(Map<String, dynamic> json) {
    id = json['id'] ?? "";
    name = json['name'] ?? "";
    type = json['type'] ?? "local";
    localPath = json['localPath'] ?? "";
  }

  FunscriptLocation();
}
